File: Cursojava22/app/src/main/java/com/example/cursojava2/MiIntentService.java
package com.example.cursojava2;

import android.app.IntentService;
import android.content.Intent;

public class MiIntentService extends IntentService {

    public static final String ACTION_PROGRESO = "com.example.cursoandroidjava.action.PROGRESO";
    public static final String ACTION_FIN = "com.example.cursoandroidjava.action.FIN";

    public MiIntentService() {
        super("MiIntentService");
    }

    @Override
    protected void onHandleIntent(Intent intent)
    {
        int iter = intent.getIntExtra("iteraciones", 0);

        for(int i=1; i<=iter; i++) {
            tareaLarga();

            //Comunicamos el progreso
            Intent bcIntent = new Intent();
            bcIntent.setAction(ACTION_PROGRESO);
            bcIntent.putExtra("progreso", i*10);
            sendBroadcast(bcIntent);
        }

        Intent bcIntent = new Intent();
        bcIntent.setAction(ACTION_FIN);
        sendBroadcast(bcIntent);
    }

    private void tareaLarga()
    {
        try {
            Thread.sleep(1000);
        } catch(InterruptedException e) {}
    }
}

File: KotlinCurso2/app/src/main/java/com/example/kotlincurso2/MiIntentService.kt
package com.example.kotlincurso2

import android.app.IntentService
import android.content.Intent


class MiIntentService : IntentService("MiIntentService") {

    override fun onHandleIntent(intent: Intent?) {
        val iter = intent!!.getIntExtra("iteraciones", 0)

        for (i in 1..iter) {
            tareaLarga()

            //Comunicamos el progreso
            val bcIntent = Intent()
            bcIntent.action = ACTION_PROGRESO
            bcIntent.putExtra("progreso", i * 10)
            sendBroadcast(bcIntent)
        }

        val bcIntent = Intent()
        bcIntent.action = ACTION_FIN
        sendBroadcast(bcIntent)
    }

    private fun tareaLarga() {
        try {
            Thread.sleep(1000)
        } catch (e: InterruptedException) {
        }

    }

    companion object {

        val ACTION_PROGRESO = "com.example.cursoandroidjava.action.PROGRESO"
        val ACTION_FIN = "com.example.cursoandroidjava.action.FIN"
    }
}
